### PR TITLE
Generalize the deprecation machinery to other forms of comments

### PR DESCRIPTION
**Pull request by herbelin**
Add a new attribute `warn(note="message", cats="cata,catb,catc")` that is automatically displayed at the time of use of the corresponding object (via the nametab, as for the deprecation) as a warning with message `message` and categories `cata`, `catb` and `catc`. This works for various kinds of objects (at the current time, mostly logical declarations and library files, but it could be extended to Ltac, Ltac2, notations, ...).

Examples:
```coq
#[warn(note="I'm a constant number", cats="constant-number-message")] Definition a := 0.
Check a.
Toplevel input, characters 6-7:
> Check a.
>       ^
Warning: I'm a constant number [warn-reference-constant-number-message,warn-reference,constant-number-message,user-warn,default]
```
In file Foo.v:
```coq
Attributes warn(note="I'm the file Foo"), deprecated(note="use Bar").
```
Then at loading time:
```coq
> Require Foo.
> ^^^^^^^^^^^^
Warning: ...
```

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.
- [x] Added / updated **documentation**.

#### Overlays (to be merged in sync with the PR)

- https://github.com/LPCIC/coq-elpi/pull/589

### CODE REVIEW

**Comment by JasonGross** [comment]
> comment-reference,default

I think we should make these comment-reference,comment,default so they can all be silenced simultaneously if desired

**Comment by proux01** [comment]
I haven't reviewed the code yet as I think we should first agree on the user facing syntax.
I still think that something like "alarm" or "warn" would be more appropriate than "information":
* it's shorter (not very important)
* it's more informative (no pun intended) considering the thing triggers a warning (as already noticed by @SkySkimmer on zulip: https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs-.26-plugin-devs/topic/Adding.20an.20.22information.22.20attribute.20beside.20.22deprecation.22/near/399916163 )
* we probably still want to add some kind of docstring in the future and having then both doc/docstring and information might become confusing

It would be nice to have an argument to the attribute enabling to set the categories, something like
```
#[information/alarm/warn/whatever(note="blah", categories="cat1,cat2,cat3")] Definition foo := ...
```
which would trigger a warning with categories `[cat1,cat2,cat3,default]`
As a side benefit, the current `deprecated(note="blah", since="XX")` attribute would just become syntactic sugar for `information/alarm/warn/whatever(note="blah", categories="deprecated,deprecated-reference,deprecated-since-XX,deprecated-reference-since-XX")` which may simplify the implementation.

**Comment by SkySkimmer** [comment]
>As a side benefit, the current deprecated(note="blah", since="XX") attribute would just become syntactic sugar for ...

Not quite, deprecated makes it say `Definition foobar is deprecated since ...` before the `note`

**Comment by proux01** [comment]
To summarize the conclusion of the discussion during [yesterday's call](https://github.com/coq/coq/wiki/Coq-Call-2023-12-04), there was an agreement on the attribute syntax `warn(note="bla", categories="cata,catb,catc")`

**Comment by ejgallego** [comment]
I'm not opposing this PR, however I am having trouble understanding the methodology here.

I see the PR has actually 0 examples of use of this definition. It seems strange to me to add annotation capabilities without first doing some research on what the actual needs of libraries and users are.

**Comment by herbelin** [comment]
@ejgallego: The typical example is #18032 (warn about vectors) where a warning is expected and not a deprecation.

**Comment by herbelin** [comment]
I rebased but this remains a basis for discussion, both about which kind of information/docstring may be added, where to store these information (currently using the same mechanism as deprecations), and how to use them (coqc, UI, ...).

**Comment by proux01** [comment]
Thanks, IIRC we reached a consensu during a weekly call in december, I'll try an implem, hopefully in the coming weeks.

**Comment by proux01** [comment]
@herbelin here is an implementation following the consensus we found during the [Call on December 4th](https://github.com/coq/coq/wiki/Coq-Call-2023-12-04).

**Comment by herbelin** [comment]
Very nice!

File: test-suite/output/deprecation_definition.v [hunk 1]
```diff
@@ -30,3 +30,9 @@ Check depr3.
 Check depr4.
 Check depr5.
 Check depr6.
+
+#[deprecated(note="deprecable"),
+  warn(note="be careful", cats="careful, be careful"),
+  warn(note="also about bla", cats="careful, careful bla")]
```
[the remark this answers is not in the document]
Comment (proux01) [reply test-suite/output/deprecation_definition.v:36]: done

File: doc/sphinx/using/libraries/writing.rst [hunk 2]
```diff
@@ -67,6 +67,43 @@ deprecated compatibility alias using :cmd:`Notation (abbreviation)`
...
+   :cmd:`Notation`, :cmd:`Infix`, :cmd:`Definition`, :cmd:`Theorem`,
+   and similar commands. To attach it to a compiled library file, use
+   :cmd:`Attributes`.
+
+   It can trigger the following warning:
+
+   .. warn:: @string__note
+
+      :n:`@string__note` is the note. It's common practice to start it
+      with a capital and end it with a dot.
```
Comment (herbelin) [inline doc/sphinx/using/libraries/writing.rst:96]: To the risk of being "pointilleux", I've been told that the American word is "period" (which Coq uses by convention) and the British word is "full stop" (but I'm not the expert).

Comment (proux01) [reply doc/sphinx/using/libraries/writing.rst]: Indeed, let's go for period.

**Comment by herbelin** [comment]
@proux01, so I guess we declare it ready!?

**Comment by proux01** [comment]
I'll merge next week if there are no more comments and CI is happy.

**Comment by proux01** [comment]
@coqbot run full ci

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
@coqbot merge now

**Pull request merged**